pkg/sidecarexec/sandbox_cmd_runner_linux.go

```diff
@@ -0,0 +1,282 @@
+		cmd.SysProcAttr.Cloneflags = cmd.SysProcAttr.Cloneflags | syscall.CLONE_NEWNET
+	}
+
+	err = r.local.Run(cmd, opts)
+	if err != nil {
+		_ = os.RemoveAll(newRootDir) // remove error is lower priority
+		return err
+	}
+
+	// Propagate up any errors deleting root directory
+	// (want to notice any deletion problems asap)
+	err = os.RemoveAll(newRootDir)
+	if err != nil {
+		return fmt.Errorf("Deleting new root dir: %s", err)
+	}
+
+	return nil
+}
+
+// RunWithCancel executes exec.Cmd.
+// Kills execution immediately if value is read from cancelCh.
+func (SandboxCmdRunner) RunWithCancel(cmd *goexec.Cmd, cancelCh chan struct{}, opts exec.RunOpts) error {
+	return fmt.Errorf("RunWithCancel not implemented")
+}
+
+type sandboxWrapArgs struct {
+	NewRootDir string
+	Posix      bool
+	Network    bool
+
+	CmdDir  string
+	CmdPath string
+	CmdArgs []string
+
+	VisiblePaths []string
+}
+
+func newSandboxWrapArgs(str string) sandboxWrapArgs {
+	var args sandboxWrapArgs
+	err := json.Unmarshal([]byte(str), &args)
+	if err != nil {
+		panic(fmt.Sprintf("Internal inconsistency: invalid sandboxWrapArgs: %s", err))
+	}
+	return args
+}
+
+func (a sandboxWrapArgs) AsString() string {
+	bs, err := json.Marshal(a)
+	if err != nil {
+		panic(fmt.Sprintf("Internal inconsistency: invalid sandboxWrapArgs: %s", err))
+	}
+	return string(bs)
+}
+
+// SandboxWrap represents continuation of sandbox setup (mounting, ...)
+// which sandbox has initiated within a new user/mount namespace.
+type SandboxWrap struct{}
+
+type mountPoint struct {
+	Path      string
+	File      bool
+	Writeable bool
+	NonMount  bool
+	Content   string
+}
+
+// ExecuteCmd mounts various directories needed to execute sandboxed command.
+// As a last step it exec-s into given command, replacing the current process.
+func (w SandboxWrap) ExecuteCmd(rawArgs []string) error {
+	runtime.LockOSThread() // as a safety precaution for syscalls below
+
+	args := newSandboxWrapArgs(rawArgs[0])
+
+	// Keep default environment as minimal as possible
+	mounts := []mountPoint{
+		{Path: "usr/bin"},
+		{Path: "tmp", NonMount: true},
+	}
+
+	for _, path := range args.VisiblePaths {
+		if !strings.HasPrefix(path, limitDir) {
+			return fmt.Errorf("Expected visible path to start with '%s', but was '%s'", limitDir, path)
+		}
+		mounts = append(mounts, mountPoint{Path: path[1:], Writeable: true})
+	}
+
+	if args.Posix {
+		// Example: calling out to git
+		mounts = append(mounts, []mountPoint{
+			{Path: "etc", NonMount: true},
+			{Path: "etc/passwd", File: true, NonMount: true, Content: "root:x:0:0:/home/kapp-controller:/usr/sbin/nologin\n"},
+			{Path: "bin"}, // e.g. bin/sh was referenced by git executables
+			{Path: "lib"},
+			{Path: "lib64"}, // some dynamic libraries
+			{Path: "var/lib"},
+			{Path: "usr/lib"},
+			{Path: "usr/libexec"}, // internal git binaries
+			{Path: "usr/share"},   // internal git data
+			{Path: "usr/sbin"},
+			// TODO get rid of 'dev/null'? (used by vendir today)
```

Comment:
just wondering why this is a goal -- is the idea just that every piece of surface area here increases risk?

Reply:
imho the more minimal the better. most things dont write things to /dev/null (or even need /dev directory). the only reason why we have it for vendir is because we run `vendir sync -f ... --lock-output /dev/null` (https://github.com/vmware-tanzu/carvel-kapp-controller/blob/develop/pkg/fetch/vendir.go#L388)